File: src/tools/registry.ts
/**
 * Tool registry for Hungarian Law MCP Server.
 * Shared between stdio (index.ts) and HTTP entry points.
 */

import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
  Tool,
} from '@modelcontextprotocol/sdk/types.js';
import type Database from '@ansvar/mcp-sqlite';

import { searchLegislation } from './search-legislation.js';
import { getProvision } from './get-provision.js';
import { validateCitationTool } from './validate-citation.js';
import { buildLegalStance } from './build-legal-stance.js';
import { formatCitationTool } from './format-citation.js';
import { checkCurrency } from './check-currency.js';
import { getEUBasis } from './get-eu-basis.js';
import { getHungarianImplementations } from './get-hungarian-implementations.js';
import { searchEUImplementations } from './search-eu-implementations.js';
import { getProvisionEUBasis } from './get-provision-eu-basis.js';
import { validateEUCompliance } from './validate-eu-compliance.js';
import { listSources } from './list-sources.js';
import { getAbout, type AboutContext } from './about.js';
export type { AboutContext } from './about.js';

const ABOUT_TOOL: Tool = {
  name: 'about',
  description:
    'Server metadata, dataset statistics, freshness, and provenance. ' +
    'Call this to verify data coverage, currency, and content basis before relying on results.',
  inputSchema: { type: 'object', properties: {}, additionalProperties: false },
  annotations: { title: 'About this server', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
};

const LIST_SOURCES_TOOL: Tool = {
  name: 'list_sources',
  description:
    'Returns detailed provenance metadata for all data sources used by this server, ' +
    'including the Nemzeti Jogszabálytár (National Legislation Database) (Magyar Közlöny (Hungarian Official Gazette)). ' +
    'Use this to understand what data is available, its authority, coverage scope, and known limitations. ' +
    'Also returns dataset statistics (document counts, provision counts) and database build timestamp. ' +
    'Call this FIRST when you need to understand what Hungarian legal data this server covers.',
  inputSchema: { type: 'object', properties: {}, additionalProperties: false },
  annotations: { title: 'List data sources', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
};

export const TOOLS: Tool[] = [
  {
    name: 'search_legislation',
    description:
      'Search Hungarian statutes and regulations by keyword using full-text search (FTS5 with BM25 ranking). ' +
      'Returns matching provisions with document context, snippets with >>> <<< markers around matched terms, and relevance scores. ' +
      'Supports FTS5 syntax: quoted phrases ("exact match"), boolean operators (AND, OR, NOT), and prefix wildcards (term*). ' +
      'Results are in English. Default limit is 10 results. For broad topics, increase the limit. ' +
      'Do NOT use this for retrieving a known provision — use get_provision instead.',
    inputSchema: {
      type: 'object',
      properties: {
        query: {
          type: 'string',
          description:
            'Search query in English. Supports FTS5 syntax: ' +
            '"personal information" for exact phrase, privacy* for prefix.',
        },
        document_id: {
          type: 'string',
          description: 'Optional: filter results to a specific statute by its document ID.',
        },
        status: {
          type: 'string',
          enum: ['in_force', 'amended', 'repealed'],
          description: 'Optional: filter by legislative status.',
        },
        limit: {
          type: 'number',
          description: 'Maximum results to return (default: 10, max: 50).',
          default: 10,
        },
      },
      required: ['query'],
    },
    annotations: { title: 'Search legislation', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'get_provision',
    description:
      'Retrieve the full text of a specific provision (section) from an Hungarian statute. ' +
      'Specify a document_id (Act title, abbreviation, or internal ID) and optionally a section or provision_ref. ' +
      'Omit section/provision_ref to get ALL provisions in the statute (use sparingly — can be large). ' +
      'Returns provision text, chapter, section number, and metadata. ' +
      'Supports Act title references (e.g., "Privacy Act 1988"), abbreviations, and full titles. ' +
      'Use this when you know WHICH provision you want. For discovery, use search_legislation instead.',
    inputSchema: {
      type: 'object',
      properties: {
        document_id: {
          type: 'string',
          description:
            'Statute identifier: Act title (e.g., "Privacy Act 1988"), abbreviation, ' +
            'or internal document ID (e.g., "privacy-act-1988").',
        },
        section: {
          type: 'string',
          description: 'Section number (e.g., "13", "8"). Omit to get all provisions.',
        },
        provision_ref: {
          type: 'string',
          description: 'Direct provision reference (e.g., "s13"). Alternative to section parameter.',
        },
      },
      required: ['document_id'],
    },
    annotations: { title: 'Get provision text', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'validate_citation',
    description:
      'Validate an Hungarian legal citation against the database — zero-hallucination check. ' +
      'Parses the citation, checks that the document and provision exist, and returns warnings about status ' +
      '(repealed, amended). Use this to verify any citation BEFORE including it in a legal analysis. ' +
      'Supports formats: "Section 13 Privacy Act 1988", "Privacy Act 1988 s 13", "s 13".',
    inputSchema: {
      type: 'object',
      properties: {
        citation: {
          type: 'string',
          description: 'Citation string to validate. Examples: "Section 13 Privacy Act 1988", "Privacy Act 1988 s 13".',
        },
      },
      required: ['citation'],
    },
    annotations: { title: 'Validate citation', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'build_legal_stance',
    description:
      'Build a comprehensive set of citations for a legal question by searching across all Hungarian statutes simultaneously. ' +
      'Returns aggregated results from multiple relevant provisions, useful for legal research on a topic. ' +
      'Use this for broad legal questions like "What are the penalties for data breaches in Hungary?" ' +
      'rather than looking up a specific known provision.',
    inputSchema: {
      type: 'object',
      properties: {
        query: {
          type: 'string',
          description: 'Legal question or topic to research (e.g., "personal information", "critical infrastructure").',
        },
        document_id: {
          type: 'string',
          description: 'Optional: limit search to one statute by document ID.',
        },
        limit: {
          type: 'number',
          description: 'Max results per category (default: 5, max: 20).',
          default: 5,
        },
      },
      required: ['query'],
    },
    annotations: { title: 'Build legal stance', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'format_citation',
    description:
      'Format an Hungarian legal citation per standard conventions. ' +
      'Three formats: "full" (formal, e.g., "Section 13, Privacy Act 1988"), ' +
      '"short" (abbreviated, e.g., "Privacy Act 1988 s 13"), "pinpoint" (section reference only, e.g., "s 13").',
    inputSchema: {
      type: 'object',
      properties: {
        citation: { type: 'string', description: 'Citation string to format.' },
        format: {
          type: 'string',
          enum: ['full', 'short', 'pinpoint'],
          description: 'Output format (default: "full").',
          default: 'full',
        },
      },
      required: ['citation'],
    },
    annotations: { title: 'Format citation', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'check_currency',
    description:
      'Check whether an Hungarian statute or provision is currently in force, amended, repealed, or not yet in force. ' +
      'Returns the document status, issued date, in-force date, and warnings. ' +
      'Essential before citing any provision — always verify currency.',
    inputSchema: {
      type: 'object',
      properties: {
        document_id: {
          type: 'string',
          description: 'Statute identifier (Act title, abbreviation, or ID).',
        },
        provision_ref: {
          type: 'string',
          description: 'Optional: provision reference to check a specific section.',
        },
      },
      required: ['document_id'],
    },
    annotations: { title: 'Check currency status', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'get_eu_basis',
    description:
      'Get the EU legal basis that an Hungarian statute references or aligns with. ' +
      'As an EU Member State, Hungary transposes EU directives and implements EU regulations ' +
      '(e.g., Privacy Act references GDPR concepts, SOCI Act aligns with NIS2 patterns). ' +
      'Returns EU document identifiers, reference types, and alignment status.',
    inputSchema: {
      type: 'object',
      properties: {
        document_id: { type: 'string', description: 'Hungarian statute identifier.' },
        include_articles: {
          type: 'boolean',
          description: 'Include specific EU article references (default: false).',
          default: false,
        },
        reference_types: {
          type: 'array',
          items: { type: 'string' },
          description: 'Optional: filter by reference type (e.g., "implements", "transposes").',
        },
      },
      required: ['document_id'],
    },
    annotations: { title: 'Get EU legal basis', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'get_hungarian_implementations',
    description:
      'Find all Hungarian statutes that reference or align with a specific EU directive or regulation. ' +
      'Given an EU document ID (e.g., "regulation:2016/679" for GDPR), returns matching Hungarian statutes. ' +
      'Note: Hungary is an EU Member State and transposes EU directives into national law.',
    inputSchema: {
      type: 'object',
      properties: {
        eu_document_id: {
          type: 'string',
          description: 'EU document ID (e.g., "regulation:2016/679" for GDPR, "directive:2022/2555" for NIS2).',
        },
        primary_only: {
          type: 'boolean',
          description: 'Return only primary referencing statutes (default: false).',
          default: false,
        },
        in_force_only: {
          type: 'boolean',
          description: 'Return only currently in-force statutes (default: false).',
          default: false,
        },
      },
      required: ['eu_document_id'],
    },
    annotations: { title: 'Find Hungarian implementations', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'search_eu_implementations',
    description:
      'Search for EU directives and regulations that are referenced by Hungarian legislation. ' +
      'Search by keyword, type (directive/regulation), or year range.',
    inputSchema: {
      type: 'object',
      properties: {
        query: { type: 'string', description: 'Keyword search across EU document titles.' },
        type: { type: 'string', enum: ['directive', 'regulation'], description: 'Filter by EU document type.' },
        year_from: { type: 'number', description: 'Filter by year (from).' },
        year_to: { type: 'number', description: 'Filter by year (to).' },
        has_hungarian_implementation: {
          type: 'boolean',
          description: 'If true, only return EU documents referenced by Hungarian legislation.',
        },
        limit: { type: 'number', description: 'Max results (default: 20, max: 100).', default: 20 },
      },
    },
    annotations: { title: 'Search EU implementations', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'get_provision_eu_basis',
    description:
      'Get the EU legal basis for a SPECIFIC provision within an Hungarian statute. ' +
      'More granular than get_eu_basis (which operates at the statute level). ' +
      'Use this for pinpoint EU alignment checks at the provision level.',
    inputSchema: {
      type: 'object',
      properties: {
        document_id: { type: 'string', description: 'Hungarian statute identifier.' },
        provision_ref: { type: 'string', description: 'Provision reference (e.g., "s13" or "13").' },
      },
      required: ['document_id', 'provision_ref'],
    },
    annotations: { title: 'Get provision EU basis', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
  {
    name: 'validate_eu_compliance',
    description:
      'Check EU alignment status for an Hungarian statute or provision. ' +
      'Detects references to EU directives, alignment status, and cross-references. ' +
      'Returns compliance status (compliant, partial, unclear, not_applicable) with warnings. ' +
      'Note: As an EU Member State, Hungary is bound by EU law. This checks transposition and compliance status.',
    inputSchema: {
      type: 'object',
      properties: {
        document_id: { type: 'string', description: 'Hungarian statute identifier.' },
        eu_document_id: { type: 'string', description: 'Optional: check against a specific EU document.' },
      },
      required: ['document_id'],
    },
    annotations: { title: 'Validate EU compliance', readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
  },
];

/** Tool name → handler. Each handler declares its own input type; raw MCP
 * args flow straight through (`never` makes every concrete input assignable). */
const HANDLERS: Record<string, (db: InstanceType<typeof Database>, args: never) => Promise<unknown>> = {
  search_legislation: searchLegislation,
  get_provision: getProvision,
  validate_citation: validateCitationTool,
  build_legal_stance: buildLegalStance,
  format_citation: formatCitationTool,
  check_currency: checkCurrency,
  get_eu_basis: getEUBasis,
  get_hungarian_implementations: getHungarianImplementations,
  search_eu_implementations: searchEUImplementations,
  get_provision_eu_basis: getProvisionEUBasis,
  validate_eu_compliance: validateEUCompliance,
  list_sources: (db) => listSources(db),
};

export function buildTools(context?: AboutContext): Tool[] {
  const tools = [...TOOLS, LIST_SOURCES_TOOL];

  if (context) {
    tools.push(ABOUT_TOOL);
  }

  return tools;
}

export function registerTools(
  server: Server,
  db: InstanceType<typeof Database>,
  context?: AboutContext,
): void {
  const allTools = buildTools(context);

  server.setRequestHandler(ListToolsRequestSchema, async () => {
    return { tools: allTools };
  });

  server.setRequestHandler(CallToolRequestSchema, async (request) => {
    const { name, arguments: args } = request.params;

    try {
      let result: unknown;

      if (name === 'about') {
        if (!context) {
          return {
            content: [{ type: 'text' as const, text: 'About tool not configured.' }],
            isError: true,
          };
        }
        result = getAbout(db, context);
      } else {
        const handler = HANDLERS[name];
        if (!handler) {
          return {
            content: [{ type: 'text' as const, text: `Error: Unknown tool "${name}".` }],
            isError: true,
          };
        }
        result = await handler(db, args as never);
      }

      return {
        content: [{ type: 'text' as const, text: JSON.stringify(result, null, 2) }],
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        content: [{ type: 'text' as const, text: `Error: ${message}` }],
        isError: true,
      };
    }
  });
}

File: internal/tools/registry.go
// Registry for Hungarian Law MCP tools — port of src/tools/registry.ts.
// Shared between the stdio and HTTP entrypoints.

package tools

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"runtime/debug"

	"github.com/google/jsonschema-go/jsonschema"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// AboutContext carries the precomputed values the about tool needs — port of
// AboutContext in src/tools/about.ts. A nil *AboutContext means the about
// tool is not configured (registered) but still answered with an explicit
// error when called.
type AboutContext struct {
	Version     string
	Fingerprint string
	DBBuilt     string
}

// Handler is the common signature of every non-about tool handler. ctx is
// the request context, so client disconnects cancel in-flight DB work; args
// is the unvalidated argument map decoded from the call payload — handlers
// parse and validate their own input. A returned error becomes an in-band
// "Error: …" tool result, never a protocol error.
type Handler func(ctx context.Context, db *sql.DB, args map[string]any) (results any, meta ResponseMetadata, err error)

// decodeCallArgs parses the raw call payload into the argument map carried
// by Handler; an empty payload is an empty map.
func decodeCallArgs(raw json.RawMessage) (map[string]any, error) {
	if len(raw) == 0 {
		return map[string]any{}, nil
	}
	var args map[string]any
	if err := json.Unmarshal(raw, &args); err != nil {
		return nil, fmt.Errorf("decode arguments: %w", err)
	}
	return args, nil
}

// decodeArgs re-encodes the raw argument map into a handler's typed args
// struct, preserving the pointer-optionality of absent keys.
func decodeArgs(args map[string]any, out any) error {
	if len(args) == 0 {
		return nil
	}
	buf, err := json.Marshal(args)
	if err != nil {
		return fmt.Errorf("encode arguments: %w", err)
	}
	if err := json.Unmarshal(buf, out); err != nil {
		return fmt.Errorf("decode arguments: %w", err)
	}
	return nil
}

// Handlers maps tool name → handler for the 12 non-about tools — port of the
// HANDLERS record in src/tools/registry.ts.
func Handlers() map[string]Handler {
	return map[string]Handler{
		"search_legislation":            SearchLegislation,
		"get_provision":                 GetProvision,
		"validate_citation":             ValidateCitation,
		"build_legal_stance":            BuildLegalStance,
		"format_citation":               FormatCitation,
		"check_currency":                CheckCurrency,
		"get_eu_basis":                  GetEUBasis,
		"get_hungarian_implementations": GetHungarianImplementations,
		"search_eu_implementations":     SearchEUImplementations,
		"get_provision_eu_basis":        GetProvisionEUBasis,
		"validate_eu_compliance":        ValidateEUCompliance,
		"list_sources": func(ctx context.Context, db *sql.DB, _ map[string]any) (any, ResponseMetadata, error) {
			return ListSources(ctx, db)
		},
	}
}

// toolDef is one entry of the registration list, in TypeScript TOOLS order.
type toolDef struct {
	name        string
	description string
	title       string
	schema      *jsonschema.Schema
}

// toolDefs lists all 13 tools in src/tools/registry.ts order
// (TOOLS + LIST_SOURCES_TOOL + ABOUT_TOOL).
func toolDefs() []toolDef {
	return []toolDef{
		{
			name: "search_legislation",
			description: "Search Hungarian statutes and regulations by keyword using full-text " +
				"search (FTS5 with BM25 ranking). " +
				"Returns matching provisions with document context, snippets with >>> <<< " +
				"markers around matched terms, and relevance scores. " +
				"Supports FTS5 syntax: quoted phrases (\"exact match\"), boolean operators " +
				"(AND, OR, NOT), and prefix wildcards (term*). " +
				"Results are in English. Default limit is 10 results. For broad topics, increase the limit. " +
				"Do NOT use this for retrieving a known provision — use get_provision instead.",
			title:  "Search legislation",
			schema: searchLegislationSchema,
		},
		{
			name: "get_provision",
			description: "Retrieve the full text of a specific provision (section) from an Hungarian statute. " +
				"Specify a document_id (Act title, abbreviation, or internal ID) and optionally a section or provision_ref. " +
				"Omit section/provision_ref to get ALL provisions in the statute (use sparingly — can be large). " +
				"Returns provision text, chapter, section number, and metadata. " +
				"Supports Act title references (e.g., \"2011. évi CXII. törvény\"), abbreviations, and full titles. " +
				"Use this when you know WHICH provision you want. For discovery, use search_legislation instead.",
			title:  "Get provision text",
			schema: getProvisionSchema,
		},
		{
			name: "validate_citation",
			description: "Validate an Hungarian legal citation against the database — zero-hallucination check. " +
				"Parses the citation, checks that the document and provision exist, and returns warnings about status " +
				"(repealed, amended). Use this to verify any citation BEFORE including it in a legal analysis. " +
				"Supports formats: \"2011. évi CXII. törvény 3. §\", \"act-cxii-2011-info-self-determination s 3\", \"s 3\".",
			title:  "Validate citation",
			schema: validateCitationSchema,
		},
		{
			name: "build_legal_stance",
			description: "Build a comprehensive set of citations for a legal question by " +
				"searching across all Hungarian statutes simultaneously. " +
				"Returns aggregated results from multiple relevant provisions, useful for legal research on a topic. " +
				"Use this for broad legal questions like \"What are the penalties for data breaches in Hungary?\" " +
				"rather than looking up a specific known provision.",
			title:  "Build legal stance",
			schema: buildLegalStanceSchema,
		},
		{
			name: "format_citation",
			description: "Format an Hungarian legal citation per standard conventions. " +
				"Two formats: \"full\" (formal, e.g., \"Infotörvény 3. §\" from \"Section 3 Infotörvény\"), " +
				"\"pinpoint\" (section reference only, e.g., \"3. §\").",
			title:  "Format citation",
			schema: formatCitationSchema,
		},
		{
			name: "check_currency",
			description: "Check whether an Hungarian statute is currently in force, amended, repealed, or not yet in force. " +
				"Returns the document status, issued date, in-force date, and warnings. " +
				"Essential before citing any provision — always verify currency.",
			title:  "Check currency status",
			schema: checkCurrencySchema,
		},
		{
			name: "get_eu_basis",
			description: "Get the EU legal basis that an Hungarian statute references or aligns with. " +
				"As an EU Member State, Hungary transposes EU directives and implements EU regulations " +
				"(e.g., Infotörvény — the Hungarian data protection act — implements GDPR). " +
				"Returns EU document identifiers, reference types, and alignment status.",
			title:  "Get EU legal basis",
			schema: getEUBasisSchema,
		},
		{
			name: "get_hungarian_implementations",
			description: "Find all Hungarian statutes that reference or align with a specific EU directive or regulation. " +
				"Given an EU document ID (e.g., \"regulation:2016/679\" for GDPR), returns matching Hungarian statutes. " +
				"Note: Hungary is an EU Member State and transposes EU directives into national law.",
			title:  "Find Hungarian implementations",
			schema: getHungarianImplementationsSchema,
		},
		{
			name: "search_eu_implementations",
			description: "Search for EU directives and regulations that are referenced by Hungarian legislation. " +
				"Search by keyword, type (directive/regulation), or year range.",
			title:  "Search EU implementations",
			schema: searchEUImplementationsSchema,
		},
		{
			name: "get_provision_eu_basis",
			description: "Get the EU legal basis for a SPECIFIC provision within an Hungarian statute. " +
				"More granular than get_eu_basis (which operates at the statute level). " +
				"Use this for pinpoint EU alignment checks at the provision level.",
			title:  "Get provision EU basis",
			schema: getProvisionEUBasisSchema,
		},
		{
			name: "validate_eu_compliance",
			description: "Check EU alignment status for an Hungarian statute or provision. " +
				"Detects references to EU directives, alignment status, and cross-references. " +
				"Returns compliance status (compliant, partial, unclear, not_applicable) with warnings. " +
				"Note: As an EU Member State, Hungary is bound by EU law. This checks transposition and compliance status.",
			title:  "Validate EU compliance",
			schema: validateEUComplianceSchema,
		},
		{
			name: "list_sources",
			description: "Returns detailed provenance metadata for all data sources used by this server, " +
				"including the Nemzeti Jogszabálytár (National Legislation Database) " +
				"(Magyar Közlöny (Hungarian Official Gazette)). " +
				"Use this to understand what data is available, its authority, coverage scope, and known limitations. " +
				"Also returns dataset statistics (document counts, provision counts) and database build timestamp. " +
				"Call this FIRST when you need to understand what Hungarian legal data this server covers.",
			title:  "List data sources",
			schema: emptyObjectSchema,
		},
		{
			name: "about",
			description: "Server metadata, dataset statistics, freshness, and provenance. " +
				"Call this to verify data coverage, currency, and content basis before relying on results.",
			title:  "About this server",
			schema: emptyObjectSchema,
		},
	}
}

// annotations mirrors the identical annotation block every TypeScript tool
// carries: { title, readOnlyHint: true, destructiveHint: false,
// idempotentHint: true, openWorldHint: false }.
func annotations(title string) *mcp.ToolAnnotations {
	return &mcp.ToolAnnotations{
		Title:           title,
		ReadOnlyHint:    true,
		DestructiveHint: new(false),
		IdempotentHint:  true,
		OpenWorldHint:   new(false),
	}
}

// Register adds all tools to the MCP server — port of registerTools in
// src/tools/registry.ts. The about tool is registered only when about is
// non-nil (mirroring the conditional ABOUT_TOOL push in buildTools), but a
// call to it is still answered with "about tool not configured" when about
// is nil, exactly as the TypeScript dispatcher does.
func Register(s *mcp.Server, db *sql.DB, about *AboutContext) {
	handlers := Handlers()
	handler := dispatch(db, about, handlers)
	for _, def := range toolDefs() {
		if def.name == "about" && about == nil {
			continue
		}
		s.AddTool(&mcp.Tool{
			Name:        def.name,
			Description: def.description,
			InputSchema: def.schema,
			Annotations: annotations(def.title),
		}, handler)
	}
}

// dispatch builds the single MCP tool-call handler that special-cases `about`
// and routes everything else through the Handlers map, passing the request
// context through so cancellation reaches the DB queries. Errors become
// in-band "Error: …" text results with IsError set — a non-nil Go error
// return would instead surface as a JSON-RPC protocol error, which the
// TypeScript server never produces.
func dispatch(db *sql.DB, about *AboutContext, handlers map[string]Handler) mcp.ToolHandler {
	return func(ctx context.Context, req *mcp.CallToolRequest) (result *mcp.CallToolResult, err error) {
		name := req.Params.Name
		// Panic isolation: the go-sdk contains no recover(), so an uncaught
		// panic would kill the stdio server with a silent EOF. Log name, value
		// and stack to stderr (never stdout — that is the protocol channel in
		// stdio mode) and answer with a generic in-band error.
		defer func() {
			if r := recover(); r != nil {
				fmt.Fprintf(os.Stderr, "panic in tool %q: %v\n%s\n", name, r, debug.Stack())
				result = errorResult("Internal tool error")
				err = nil
			}
		}()

		var (
			results    any
			meta       ResponseMetadata
			handlerErr error
		)
		if name == "about" {
			if about == nil {
				return errorResult(errAboutNotConfigured.Error()), nil
			}
			results, meta, handlerErr = GetAbout(ctx, db, about)
		} else {
			h, ok := handlers[name]
			if !ok {
				return errorResult(fmt.Sprintf("Error: Unknown tool %q.", name)), nil
			}
			var args map[string]any
			args, handlerErr = decodeCallArgs(req.Params.Arguments)
			if handlerErr == nil {
				results, meta, handlerErr = h(ctx, db, args)
			}
		}
		if handlerErr != nil {
			// Server-side record: in-band results are invisible to the
			// operator, so a broken DB would otherwise go unnoticed (stderr is
			// safe in both transports — stdout is the protocol channel in
			// stdio mode).
			fmt.Fprintf(os.Stderr, "tool %q failed: %v\n", name, handlerErr)
			return errorResult("Error: " + handlerErr.Error()), nil
		}
		text := MarshalResponse(results, meta)
		if name == "about" {
			// TypeScript stringifies getAbout's bare return object directly
			// (registry.ts:378) — the only tool without the results/_metadata
			// envelope.
			text = MarshalBare(results)
		}
		return &mcp.CallToolResult{
			Content: []mcp.Content{&mcp.TextContent{Text: text}},
		}, nil
	}
}

// errorResult builds the { content: [{type:'text', text}], isError: true }
// shape the TypeScript registry returns for failures.
func errorResult(text string) *mcp.CallToolResult {
	return &mcp.CallToolResult{
		Content: []mcp.Content{&mcp.TextContent{Text: text}},
		IsError: true,
	}
}
